app/Http/Controllers/SquareCheckoutController.php

```diff
@@ -0,0 +1,221 @@
+        ]);
+
+        $checkoutApi = $square->getCheckoutApi();
+        $checkoutResponse = $checkoutApi->createCheckout(config('square.location_id'), $checkoutRequest);
+
+        if (! $checkoutResponse->isSuccess()) {
+            Log::error(self::class.' Error creating checkout - '.json_encode($checkoutResponse->getErrors()));
+
+            return view(
+                'square.error',
+                [
+                    'message' => 'We could not contact Square to begin your payment.',
+                ]
+            );
+        }
+
+        $checkout = $checkoutResponse->getResult()->getCheckout();
+
+        $payment->checkout_id = $checkout->getId();
+        $payment->order_id = $checkout->getOrder()->getId();
+        $payment->save();
+
+        return redirect($checkout->getCheckoutPageUrl());
+    }
+
+    public function complete(SquareCompleteRequest $request)
+    {
+        $payment = Payment::where('id', $request->input('referenceId'))->firstOrFail();
+
+        if ($payment->checkout_id !== $request->input('checkoutId')) {
+            return view(
+                'square.error',
+                [
+                    'message' => 'We could not match your payment in our database.',
+                ]
+            );
+        }
+
+        $square = new SquareClient([
+            'accessToken' => config('square.access_token'),
+            'environment' => config('square.environment'),
+        ]);
+
+        $ordersApi = $square->getOrdersApi();
+
+        $retrieveOrderResponse = $ordersApi->retrieveOrder($payment->order_id);
+
+        if (! $retrieveOrderResponse->isSuccess()) {
+            Log::error(self::class.' Error retrieving order - '.json_encode($retrieveOrderResponse->getErrors()));
+
+            return view(
+                'square.error',
+                [
+                    'message' => 'We could not contact Square to finalize your payment.',
+                ]
+            );
+        }
+
+        $order = $retrieveOrderResponse->getResult()->getOrder();
+
+        switch ($order->getState()) {
+            case OrderState::COMPLETED:
+                $tender = $order->getTenders()[0];
+                $payment->amount = $tender->getAmountMoney()->getAmount() / 100;
+
+                $processingFeeMoney = $tender->getProcessingFeeMoney();
+                if (null !== $processingFeeMoney) {
+                    $payment->processing_fee = $processingFeeMoney->getAmount() / 100;
+                }
+
+                $cardDetails = $tender->getCardDetails();
+                if (null !== $cardDetails) {
+                    $card = $cardDetails->getCard();
+                    if (null !== $card) {
+                        $payment->card_brand = $card->getCardBrand();
+                        $payment->card_type = $card->getCardType();
+                        $payment->last_4 = $card->getLast4();
+                        $payment->prepaid_type = $card->getPrepaidType();
```

Comment:
What is this?

Reply:
Flag from Square that indicates whether this is a prepaid card or not.